File: hooks/mechanical-floor-gate.py
#!/usr/bin/env -S uv run python3
"""PreToolUse gate: guarantee the deterministic static-analysis floor (check-all) actually RAN and
PASSED before a phase spends tokens on its expensive fan-out. Prose "run check-all" is a suggestion
the model can skip; this is the enforcement (the plugin's "Hooks over prompt" doctrine). This is the
dev/ds sibling of hooks/writing-mechanical-gate.py — the same gap (the Leg-1 mechanical floor sat in
skippable prose while the phase hook gated on a *different* artifact) the wc-audit P20 sub-probe
flagged across writing, dev, and ds.

Parameterized by the FLOOR env var (set in the skill frontmatter hook command):

  FLOOR=dev  → gate the Agent spawn (the goal-backward verifier in dev-verify, Leg 2).
               Runs references/constraints/check-all.py, which self-scopes to the dev constraints via
               its APPLIES_TO + detected-workflow filter. Denies on HARD failures only.
               Fixes are ordinary main-chat edits, so denying Agent does not deadlock.

  FLOOR=ds   → gate the Workflow spawn (the ds-validate-coverage per-requirement fan-out).
               Runs scripts/check-all-ds.sh; denies on non-zero exit.
               Gates the WORKFLOW only — NOT Agent — on purpose: ds forbids main-chat analysis-code
               edits (ds-no-main-chat-code-guard), so a failing floor is fixed by a FIX SUBAGENT
               (an Agent). Denying Agent here would wall off the only legal way to fix it. The
               coverage Workflow is the expensive fan-out worth protecting; fix subagents stay free.

TIGHTLY SCOPED on purpose (the "check-all runs all .py" history): this hook is wired only into
dev-verify (FLOOR=dev) and ds-validate (FLOOR=ds) frontmatter, and each branch gates exactly one
tool. It never fires on Write/Edit and never drags in another workflow's constraints (check-all
self-scopes by detected workflow / the ds-*.py glob).

In all cases a constraint that ERRORED (threw — e.g. a missing dependency) is surfaced but does NOT
block: a broken/under-provisioned constraint must not wall off the phase. Only real content
FAILURES block. A harness error running the gate itself never blocks either (fail-open).

CLI (debug):  FLOOR=dev uv run python3 mechanical-floor-gate.py /abs/project
              FLOOR=ds  uv run python3 mechanical-floor-gate.py /abs/project
"""
import json
import os
import subprocess
import sys
from pathlib import Path

HOOKS_DIR = Path(__file__).resolve().parent
REPO = HOOKS_DIR.parent
CHECK_ALL_PY = REPO / "references" / "constraints" / "check-all.py"
CHECK_ALL_DS = REPO / "scripts" / "check-all-ds.sh"


def _run_dev(project: str):
    """check-all.py (dev). Return (ok, failed, errors, summary). ok iff no hard content FAILURES.

    Run with `--with lxml` so the lxml-dependent constraints actually run instead of landing in
    `errors` (the same invocation fix shipped in v5.67.5)."""
    try:
        out = subprocess.run(
            ["uv", "run", "--with", "lxml", "python3", str(CHECK_ALL_PY), project],
            capture_output=True, text=True, timeout=180)
    except Exception as e:
        return True, [], [], f"(check-all.py could not run: {e})"
    failed, errors = [], []
    try:
        raw = out.stdout
        data = json.loads(raw[:raw.rfind("}") + 1])  # strip the trailing human summary line
        failed = [f.get("name", "?") if isinstance(f, dict) else str(f) for f in data.get("failed", [])]
        errors = [e.get("name", "?") if isinstance(e, dict) else str(e) for e in data.get("errors", [])]
    except Exception:
        if out.returncode != 0:
            failed = [(out.stdout.strip().splitlines() or ["check-all reported failures"])[-1]]
    summary = (out.stdout.strip().splitlines() or ["(no output)"])[-1]
    return (len(failed) == 0), failed, errors, summary


def _run_ds(project: str):
    """check-all-ds.sh (ds). Return (ok, failed, errors, summary). ok iff exit 0.

    The bash runner prints `✗ <name>` per failing constraint and exits non-zero iff any failed."""
    try:
        out = subprocess.run(
            ["bash", str(CHECK_ALL_DS), project],
            capture_output=True, text=True, timeout=180)
    except Exception as e:
        return True, [], [], f"(check-all-ds.sh could not run: {e})"
    failed = [ln.split("✗", 1)[1].strip() for ln in out.stdout.splitlines() if "✗" in ln]
    summary = (out.stdout.strip().splitlines() or ["(no output)"])[-1]
    ok = out.returncode == 0
    if not ok and not failed:
        failed = [summary]
    return ok, failed, [], summary


def deny(reason: str):
    print(json.dumps({"hookSpecificOutput": {
        "hookEventName": "PreToolUse",
        "permissionDecision": "deny",
        "permissionDecisionReason": reason,
    }}))
    sys.exit(0)


def _project_from_args(tool_input: dict) -> str:
    args = tool_input.get("args")
    if isinstance(args, str):
        try:
            args = json.loads(args)
        except Exception:
            args = {}
    if isinstance(args, dict) and args.get("projectDir"):
        return str(args["projectDir"])
    return "."


def main():
    floor = os.environ.get("FLOOR", "").strip().lower()

    # CLI debug mode
    if len(sys.argv) > 1 and sys.argv[1] not in ("-",):
        runner = _run_ds if floor == "ds" else _run_dev
        ok, failed, errors, summary = runner(sys.argv[1])
        print(f"FLOOR={floor or 'dev'} ok={ok} | {summary}")
        if failed:
            print("FAILED (blocking):\n- " + "\n- ".join(failed))
        if errors:
            print("errors (non-blocking — tooling):\n- " + "\n- ".join(errors))
        sys.exit(0 if ok else 1)

    try:
        hook_input = json.load(sys.stdin)
    except Exception:
        sys.exit(0)
    tool_name = hook_input.get("tool_name", "")
    tool_input = hook_input.get("tool_input", {}) or {}

    if floor == "ds":
        # Gate the ds-validate-coverage Workflow fan-out ONLY (never Agent — fix subagents must run).
        if tool_name != "Workflow":
            sys.exit(0)
        project = _project_from_args(tool_input)
        ok, failed, errors, summary = _run_ds(project)
        if not ok:
            deny(
                "GATE BLOCKED: the DS static-analysis floor (check-all-ds.sh — determinism, join "
                "audits, idempotency, error handling, schema contracts, standard errors, viz "
                "integrity) has failures, so the per-requirement validation fan-out must not run "
                "yet. These are code-quality defects in the analysis scripts. Dispatch a FIX "
                "SUBAGENT (an Agent — not blocked) to fix them, then re-invoke:\n- "
                + "\n- ".join(failed or [summary])
                + "\n\n(Run `bash scripts/check-all-ds.sh .` for details.)")
        sys.exit(0)

    # FLOOR=dev (default): gate the goal-backward verifier Agent spawn.
    if tool_name != "Agent":
        sys.exit(0)
    project = _project_from_args(tool_input)
    ok, failed, errors, summary = _run_dev(project)
    if not ok:
        note = ("\n\n(Plus " + str(len(errors)) + " constraint(s) errored — tooling, NOT blocking.)"
                if errors else "")
        deny(
            "GATE BLOCKED: the constraint floor (check-all.py — Leg 1) has hard failures, so the "
            "goal-backward verifier must not run yet. Constraint failures are hard blocks — fix "
            "these first, then re-spawn the verifier:\n- "
            + "\n- ".join(failed or [summary]) + note
            + "\n\n(Run `uv run --with lxml python3 references/constraints/check-all.py .` for details.)")
    sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/mechanical-floor-gate.ts
#!/usr/bin/env bun
/**
 * PreToolUse gate: guarantee the deterministic static-analysis floor (check-all) actually RAN and
 * PASSED before a phase spends tokens on its expensive fan-out. Prose "run check-all" is a suggestion
 * the model can skip; this is the enforcement (the plugin's "Hooks over prompt" doctrine).
 *
 * TypeScript port of hooks/mechanical-floor-gate.py — behavior-preserving, including the odd bits
 * (Python's `str(bool)` capitalization in the CLI summary, the `raw[:raw.rfind("}")+1]` slice, the
 * fail-open bare excepts).
 *
 * Parameterized by the FLOOR env var (set in the skill frontmatter hook command):
 *
 *   FLOOR=dev  → gate the Agent spawn (the goal-backward verifier in dev-verify, Leg 2).
 *                Runs references/constraints/check-all.py. Denies on HARD failures only.
 *
 *   FLOOR=ds   → gate the shared beat-implement Workflow spawn.
 *                Runs scripts/check-all-ds.sh; denies on non-zero exit. Gates the WORKFLOW only —
 *                NOT Agent — so fix subagents stay free.
 *
 * A constraint that ERRORED is surfaced but does NOT block. A harness error running the gate itself
 * never blocks either (fail-open).
 *
 * CLI (debug):  FLOOR=dev bun mechanical-floor-gate.ts /abs/project
 */
import { resolve, dirname, join } from "node:path";
import { deny, denyOnCrash, parsePayload, projectFromArgs } from "./_gate_common.ts";

// FIRST STATEMENT WITH AN EFFECT: a throw below becomes a schema-valid deny instead of an
// exit-1, which Claude Code treats as NON-BLOCKING — i.e. a silent allow in a PreToolUse gate.
denyOnCrash("MECHANICAL FLOOR GATE");

const HOOKS_DIR = resolve(import.meta.dir);
const REPO = dirname(HOOKS_DIR);
const CHECK_ALL_PY = join(REPO, "references", "constraints", "check-all.py");
const CHECK_ALL_DS = join(REPO, "scripts", "check-all-ds.sh");

/** Python's `str.splitlines()`: no trailing empty element, and "" -> []. */
function pySplitlines(s: string): string[] {
  if (s === "") return [];
  const parts = s.split(/\r\n|[\n\r\v\f\x1c\x1d\x1e\x85\u2028\u2029]/);
  if (parts.length > 1 && parts[parts.length - 1] === "") parts.pop();
  return parts;
}

/** Python's `str.strip()` (whitespace both ends) followed by `splitlines()`. */
function strippedLines(s: string): string[] {
  return pySplitlines(s.trim());
}

/** `(lines or [fallback])[-1]` */
function lastLine(s: string, fallback: string): string {
  const lines = strippedLines(s);
  return lines.length ? lines[lines.length - 1] : fallback;
}

type RunOut = { stdout: string; returncode: number };

/** subprocess.run(..., capture_output=True, text=True, timeout=180). Throws like Python does. */
function runProc(cmd: string[]): RunOut {
  const r = Bun.spawnSync(cmd, { stdout: "pipe", stderr: "pipe", timeout: 180_000 });
  if (r.signalCode === "SIGTERM" || r.signalCode === "SIGKILL") throw new Error("timeout");
  const stdout = new TextDecoder().decode(r.stdout ?? new Uint8Array()).replace(/\r\n/g, "\n");
  return { stdout, returncode: r.exitCode ?? 1 };
}

type FloorResult = { ok: boolean; failed: string[]; errors: string[]; summary: string };

/** check-all.py (dev). ok iff no hard content FAILURES. */
function runDev(project: string): FloorResult {
  let out: RunOut;
  try {
    out = runProc(["uv", "run", "--with", "lxml", "python3", CHECK_ALL_PY, project]);
  } catch (e) {
    return { ok: true, failed: [], errors: [], summary: `(check-all.py could not run: ${e})` };
  }
  let failed: string[] = [];
  let errors: string[] = [];
  try {
    const raw = out.stdout;
    // strip the trailing human summary line — mirrors raw[:raw.rfind("}") + 1] exactly,
    // including the rfind == -1 case that slices to "" and makes the parse throw.
    const data = JSON.parse(raw.slice(0, raw.lastIndexOf("}") + 1));
    if (data === null || typeof data !== "object" || Array.isArray(data)) throw new Error("not a dict");
    const name = (f: unknown) =>
      f !== null && typeof f === "object" && !Array.isArray(f)
        ? String((f as Record<string, unknown>).name ?? "?")
        : String(f);
    failed = ((data.failed ?? []) as unknown[]).map(name);
    errors = ((data.errors ?? []) as unknown[]).map(name);
  } catch {
    if (out.returncode !== 0) {
      failed = [lastLine(out.stdout, "check-all reported failures")];
    }
  }
  const summary = lastLine(out.stdout, "(no output)");
  return { ok: failed.length === 0, failed, errors, summary };
}

/** check-all-ds.sh (ds). ok iff exit 0. */
function runDs(project: string): FloorResult {
  let out: RunOut;
  try {
    out = runProc(["bash", CHECK_ALL_DS, project]);
  } catch (e) {
    return { ok: true, failed: [], errors: [], summary: `(check-all-ds.sh could not run: ${e})` };
  }
  let failed = pySplitlines(out.stdout)
    .filter((ln) => ln.includes("✗"))
    .map((ln) => ln.slice(ln.indexOf("✗") + 1).trim());
  const summary = lastLine(out.stdout, "(no output)");
  const ok = out.returncode === 0;
  if (!ok && failed.length === 0) failed = [summary];
  return { ok, failed, errors: [], summary };
}

async function main(): Promise<never> {
  const floor = (process.env.FLOOR ?? "").trim().toLowerCase();
  const argv = process.argv.slice(2);

  // CLI debug mode
  if (argv.length > 0 && argv[0] !== "-") {
    const runner = floor === "ds" ? runDs : runDev;
    const { ok, failed, errors, summary } = runner(argv[0]);
    console.log(`FLOOR=${floor || "dev"} ok=${ok ? "True" : "False"} | ${summary}`);
    if (failed.length) console.log("FAILED (blocking):\n- " + failed.join("\n- "));
    if (errors.length) console.log("errors (non-blocking — tooling):\n- " + errors.join("\n- "));
    process.exit(ok ? 0 : 1);
  }

  // A PreToolUse GATE DENIES ON A PAYLOAD IT CANNOT READ. The `catch { exit 0 }` here was
  // Python parity, and it is precisely what `denyOnCrash` cannot reach: the handler covers
  // throws that ESCAPE, and a local catch means none does. Measured — unparseable stdin, and
  // for the raw-`JSON.parse` gates also `null`/`"s"`/`[1,2]`, produced exit 0 with no output,
  // i.e. a silent ALLOW on every malformed payload. `parsePayload` denies on a non-object and
  // lets a parse error propagate to the handler, which denies too.
  const hookInput: Record<string, unknown> = parsePayload(await Bun.stdin.text());
  const toolName = String(hookInput.tool_name ?? "");
  const toolInput = (hookInput.tool_input as Record<string, unknown>) ?? {};

  if (floor === "ds") {
    // Gate the shared DS implementation Workflow ONLY (never Agent — fix subagents must run).
    if (toolName !== "Workflow") process.exit(0);
    const project = projectFromArgs(toolInput, hookInput);
    const { ok, failed, summary } = runDs(project);
    if (!ok) {
      deny(
        "GATE BLOCKED: the DS static-analysis floor (check-all-ds.sh — determinism, join " +
          "audits, idempotency, error handling, schema contracts, standard errors, viz " +
          "integrity) has failures, so the shared implementation fan-out must not run " +
          "yet. These are code-quality defects in the analysis scripts. Dispatch a FIX " +
          "SUBAGENT (an Agent — not blocked) to fix them, then re-invoke:\n- " +
          (failed.length ? failed : [summary]).join("\n- ") +
          "\n\n(Run `bash scripts/check-all-ds.sh .` for details.)",
      );
    }
    process.exit(0);
  }

  // FLOOR=dev (default): gate the goal-backward verifier Agent spawn.
  if (toolName !== "Agent") process.exit(0);
  const project = projectFromArgs(toolInput, hookInput);
  const { ok, failed, errors, summary } = runDev(project);
  if (!ok) {
    const note = errors.length
      ? "\n\n(Plus " + String(errors.length) + " constraint(s) errored — tooling, NOT blocking.)"
      : "";
    deny(
      "GATE BLOCKED: the constraint floor (check-all.py — Leg 1) has hard failures, so the " +
        "goal-backward verifier must not run yet. Constraint failures are hard blocks — fix " +
        "these first, then re-spawn the verifier:\n- " +
        (failed.length ? failed : [summary]).join("\n- ") +
        note +
        "\n\n(Run `uv run --with lxml python3 references/constraints/check-all.py .` for details.)",
    );
  }
  process.exit(0);
}

await main();
